apollo-router/src/query_planner/fetch.rs

```diff
@@ -243,6 +255,189 @@ impl Display for QueryHash {
 pub(crate) struct Variables {
     pub(crate) variables: Object,
     pub(crate) inverted_paths: Vec<Vec<Path>>,
+    pub(crate) contextual_args: Option<(HashSet<String>, usize)>,
+}
+
+#[derive(Debug)]
+enum ContextBatchingError {
+    InvalidDocument(WithErrors<Document>),
+    NoSelectionSet,
+}
+
+fn query_batching_for_contextual_args(
+    operation: &str,
+    contextual_args: &Option<(HashSet<String>, usize)>,
+) -> Result<Option<String>, ContextBatchingError> {
+    if let Some((ctx, times)) = contextual_args {
+        let parser = apollo_compiler::Parser::new()
+            .parse_ast(operation, "")
+            .map_err(ContextBatchingError::InvalidDocument)?;
+        if let Some(mut operation) = parser
+            .definitions
+            .into_iter()
+            .find_map(|definition| definition.as_operation_definition().cloned())
+        {
+            let mut new_variables: Vec<_> = Default::default();
+            if operation
+                .variables
+                .iter()
+                .any(|v| ctx.contains(v.name.as_str()))
+            {
+                let new_selection_set: Vec<_> = (0..*times)
+                    .map(|i| {
+                        // TODO: Unwrap
+                        let mut s = operation
+                            .selection_set
+                            .first()
+                            .ok_or_else(|| ContextBatchingError::NoSelectionSet)?
+                            .clone();
+                        if let ast::Selection::Field(f) = &mut s {
+                            let f = f.make_mut();
+                            f.alias = Some(Name::new_unchecked(format!("_{}", i).into()));
+                        }
+
+                        for v in &operation.variables {
+                            if ctx.contains(v.name.as_str()) {
+                                let mut cloned = v.clone();
+                                let new_variable = cloned.make_mut();
+                                new_variable.name =
+                                    Name::new_unchecked(format!("{}_{}", v.name, i).into());
+                                new_variables.push(Node::new(new_variable.clone()));
+
+                                s = rename_variables(s, v.name.clone(), new_variable.name.clone());
+                            } else if !new_variables.iter().any(|var| var.name == v.name) {
+                                new_variables.push(v.clone());
+                            }
+                        }
+
+                        Ok(s)
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                let new_operation = operation.make_mut();
+                new_operation.selection_set = new_selection_set;
+                new_operation.variables = new_variables;
+
```

Comment:
Is it possible to have generated an invalid operation at this point? Do we need to re-validate here?

Reply:
I would say it's unlikely.